File: tools/doc_packs.py
"""Build the datasheet block every corpus board would get in the browser.

    .venv/Scripts/python.exe tools/doc_packs.py
    .venv/Scripts/python.exe -m harness.run --detector v8 --trials 3 \
        --documentation results/doc-packs-after.json

The page can attach a datasheet to a part; the harness never could, so five
trials said nothing about the pack a visitor with documents actually gets. This
closes that: the corpus is built here, handed to the browser's own retrieval,
and the resulting block is written out per case for `--documentation` to read.

Retrieval stays in `site/rag.js`. Four things in this repository are written
twice and each costs a parity test - a fifth would cost another, for a path that
only ever runs in a browser.

Two files come out, `before` and `after`, so the benchmark has both sides: three
whole chunks per part with no ceiling, against the budget that replaced it.
"""

from __future__ import annotations

import json
import subprocess
import sys
import tempfile
from pathlib import Path

ROOT = Path(__file__).resolve().parent.parent
sys.path.insert(0, str(ROOT))

from console import utf8  # noqa: E402
from harness.research import _PAGE_BREAK, fetch, major_parts, text_of  # noqa: E402
from harness.run import BOARDS, corpus  # noqa: E402

RESULTS = ROOT / "results"


def pages_of_cached_datasheets() -> dict[str, dict]:
    """Every cached datasheet as one string per page, keyed by the ref.

    Offline, always. A benchmark that depends on a vendor CDN being up is not a
    benchmark, and the cache is populated deliberately by `harness.research`.

    Every page, not the first twelve `harness.research` reads for its regexes:
    the browser indexes the whole document, and a budget measured against a
    truncated one would be measuring the wrong thing.
    """
    out: dict[str, dict] = {}
    for name, path in BOARDS.items():
        board = json.loads(path.read_text(encoding="utf-8"))
        for part in major_parts(board):
            ref = part["ref"]
            if ref in out:
                continue
            url = part["datasheet"].strip()
            pdf = fetch(url, offline=True)
            if pdf is None:
                print(f"  {ref:<4} {part['value']:<22} not cached, skipped")
                continue
            pages = [" ".join(p.split()) for p in text_of(pdf, pages=999).split(_PAGE_BREAK)]
            out[ref] = {
                "part": part["value"],
                "source": url.rsplit("/", 1)[-1],
                "pages": pages,
            }
            words = sum(len(p.split()) for p in pages)
            print(f"  {ref:<4} {part['value']:<22} {len(pages):>3} pages, {words:>6} words")
    return out


def main() -> int:
    utf8()
    print("cached datasheets")
    pages = pages_of_cached_datasheets()
    if not pages:
        raise SystemExit(
            "no datasheets cached. Run `python -m harness.research` once, online, "
            "to populate .cache/datasheets."
        )

    cases = [{"id": c["id"], "board": c["board"]} for c in corpus()]
    print(f"\n{len(cases)} corpus cases\n")

    with tempfile.TemporaryDirectory() as tmp:
        tmp = Path(tmp)
        (tmp / "cases.json").write_text(json.dumps(cases), encoding="utf-8")
        (tmp / "pages.json").write_text(json.dumps(pages), encoding="utf-8")
        both = tmp / "both.json"
        built = subprocess.run(
            ["node", str(ROOT / "tools" / "doc_packs.mjs"), str(tmp / "cases.json"),
             str(tmp / "pages.json"), str(both)],
            check=True,
            capture_output=True,
            text=True,
            encoding="utf-8",
        )
        print(built.stdout)
        blocks = json.loads(both.read_text(encoding="utf-8"))

    RESULTS.mkdir(parents=True, exist_ok=True)
    for variant in ("before", "after"):
        out = RESULTS / f"doc-packs-{variant}.json"
        out.write_text(json.dumps(blocks[variant], indent=1), encoding="utf-8")
        print(f"wrote {out}")
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: tools/doc_packs.mjs
// The datasheet block each corpus board would get in the browser.
//
// The page attaches documents to parts and retrieves passages from them; the
// harness runs offline and never has. That gap is why the browser's pack went
// unmeasured long enough to grow to four fifths datasheet text - so this hands
// the browser's own retrieval across to the sweep rather than reimplementing it
// in Python, which would make retrieval the fifth thing written twice.
//
// Two variants, because the point is a comparison:
//
//   before  three whole chunks per documented part, no ceiling. What the page
//           shipped, and what made the board 21% of its own pack.
//   after   passagesBlock's budget: a share of the board, spent round-robin,
//           each passage narrowed to the span that earned it.
//
//   node tools/doc_packs.mjs <cases.json> <pages.json> <out.json>

import { readFileSync, writeFileSync } from "node:fs";
import { fileURLToPath } from "node:url";
import { dirname, join } from "node:path";

const root = join(dirname(fileURLToPath(import.meta.url)), "..");
const docs = await import("file://" + join(root, "site", "docs.js"));
const { pinsByRef } = await import("file://" + join(root, "site", "checks.js"));
const { distill } = await import("file://" + join(root, "site", "distill.js"));

const [casesPath, pagesPath, outPath] = process.argv.slice(2);
const cases = JSON.parse(readFileSync(casesPath, "utf8"));
const pages = JSON.parse(readFileSync(pagesPath, "utf8"));

const wordsIn = (s) => String(s || "").split(/\s+/).filter(Boolean).length;

/**
 * The section as it stood before the budget: three whole chunks per part.
 *
 * Kept here rather than in site/, because it is the thing being argued against
 * and a benchmark needs both sides. It is the old `passagesBlock` verbatim -
 * `passagesFor(..., 3)` with no width, which is a whole 350-word chunk.
 */
function unbudgeted(board) {
  const attached = docs.docsFor(board.meta.name);
  if (!attached.size) return "";
  const pins = pinsByRef(board);
  const lines = ["RETRIEVED DOCUMENTATION  verbatim, from documents attached to this board"];
  for (const ref of [...attached.keys()].sort()) {
    for (const hit of docs.passagesFor(board, ref, pins.get(ref) || [], 3)) {
      const where = [`${hit.source} p${hit.page}`, hit.section].filter(Boolean).join(", ");
      lines.push(`${ref} (${where}): ${hit.text}`);
    }
  }
  return lines.length > 1 ? lines.join("\n") : "";
}

/** What the page builds now: budgeted at a share of the distilled board. */
const RATIO = 0.35;
function budgeted(board) {
  const block = docs.passagesBlock(board, Math.round(wordsIn(distill(board)) * RATIO));
  return block.length ? block.join("\n") : "";
}

const out = { before: {}, after: {} };
const rows = [];

for (const item of cases) {
  const board = item.board;
  // A fresh store per case. `loadDocs` clears the built indexes too, which
  // matters: an index keyed on board name and ref would otherwise be reused
  // across two boards of the same name carrying different edits.
  await docs.loadDocs(board.meta.name);
  let attached = 0;
  for (const [ref, doc] of Object.entries(pages)) {
    // Designator AND part. Both boards have a U1 and they are different chips,
    // so matching on the designator alone handed dcdcc's regulator the
    // AMS1117's datasheet - the same mistake as joining a schematic to a layout
    // by reference, which this repository already learned the hard way.
    if (!board.components.some((c) => c.ref === ref && c.value === doc.part)) continue;
    await docs.attach(board.meta.name, ref, { name: doc.source, pages: doc.pages });
    attached += 1;
  }
  const before = unbudgeted(board);
  const after = budgeted(board);
  out.before[item.id] = before;
  out.after[item.id] = after;

  const boardWords = wordsIn(distill(board));
  rows.push({
    id: item.id,
    attached,
    board: boardWords,
    before: wordsIn(before),
    after: wordsIn(after),
  });
}

writeFileSync(outPath, JSON.stringify(out, null, 1), "utf8");

const pad = (s, n) => String(s).padEnd(n);
const num = (s, n) => String(s).padStart(n);
console.log(
  `${pad("case", 24)}${num("docs", 5)}${num("board", 7)}${num("before", 8)}${num("sheet%", 8)}` +
    `${num("after", 8)}${num("sheet%", 8)}`
);
console.log("-".repeat(68));
for (const r of rows) {
  const bp = r.before ? Math.round((r.before / (r.board + r.before)) * 100) : 0;
  const ap = r.after ? Math.round((r.after / (r.board + r.after)) * 100) : 0;
  console.log(
    pad(r.id, 24) + num(r.attached, 5) + num(r.board, 7) +
      num(r.before, 8) + num(bp + "%", 8) + num(r.after, 8) + num(ap + "%", 8)
  );
}
console.log(`\nwrote ${outPath}`);
